### ISSUE TITLE
Hazelcast 3.7.x client cannot obtain cache proxy from 3.6.x cluster

### ISSUE THREAD

**Issue by vbekiaris** [hazelcast/hazelcast#9004]
Steps to reproduce:
- Start a 3.6.x cluster with a `Cache` configured
- Start a 3.7.x client, then obtain a reference to the configured cache:

```
HazelcastInstance hz = HazelcastClient.newHazelcastClient();
CachingProvider cachingProvider = HazelcastClientCachingProvider.createCachingProvider(hz);
CacheManager cacheManager = cachingProvider.getCacheManager();
Cache<String,String> cache = cacheManager.getCache("test", String.class, String.class);
```

Fails with exception:

```
Exception in thread "main" com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: 0, id: 0, class: 'com.hazelcast.config.CacheEvictionConfig', exception: java.lang.Class cannot be cast to com.hazelcast.internal.eviction.EvictionPolicyComparator
    at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.rethrowReadException(DataSerializableSerializer.java:141)
    at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:130)
    at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:52)
    at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:216)
    at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:600)
    at com.hazelcast.config.CacheConfig.readData(CacheConfig.java:566)
    at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:127)
    at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:52)
    at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:172)
    at com.hazelcast.client.cache.impl.ClientCacheHelper.getCacheConfig(ClientCacheHelper.java:69)
    at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.getCacheConfig(HazelcastClientCacheManager.java:107)
    at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.findCacheConfig(HazelcastClientCacheManager.java:131)
    at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCacheUnchecked(AbstractHazelcastCacheManager.java:245)
    at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCache(AbstractHazelcastCacheManager.java:186)
    at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCache(AbstractHazelcastCacheManager.java:67)
    at CacheClientBootstrap.main(CacheClientBootstrap.java:35)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.lang.ClassCastException: java.lang.Class cannot be cast to com.hazelcast.internal.eviction.EvictionPolicyComparator
    at com.hazelcast.config.EvictionConfig.readData(EvictionConfig.java:264)
    at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:127)
    ... 21 more
```

**Comment by ihsandemir** [comment]
We discussed a solution for this. The client will send the request encoded as in 3.6.x version. To do this, the Authentication response shall be extended to have an extra field, the serverVersion. If serverVersion exists, then use the new CacheConfig class (as it is in 3.7) otherwise it shall use the legacy CacheConfig class.

**Comment by jerrinot** [comment]
fixed by #9066
